Create rule S6372:  "await_suspend" should accept type-erased "coroutine_handle" when it uses it in a generic way (#499)
SonarSource/rspec: rules/S6372/cfamily/rule.adoc

The implementation of the `await_suspend` method accepts the handle to the suspended coroutine as the parameter.
This parameter can be defined with either specific promise type `coroutine_handle<PromiseType>` or type erased `coroutine_handle<>`.

> The former allows `await_suspend` to access the promise of the coroutine, however, it ties the implementation to a particular type.

Reviewer: ```suggestion
The former allows `await_suspend` to access the promise of the coroutine; however, it ties the implementation to a particular type.
```

> The implement of the `await_suspend` method accepts the handle to suspended coroutine as the parameter.

Reviewer 2: ```suggestion
The implementation of the `await_suspend` method accepts the handle to the suspended coroutine as the parameter.
```

This parameter can be defined with either specific promise type `coroutine_handle<PromiseType>` or type erased `coroutine_handle<>`.
The former allows `await_suspend` to access the promise of the coroutine, however, it ties the implementation to a particular type.

> In a contrast `coroutine_handle<>` implementation supports all promise types.

Reviewer 2: I would add/modify this sentence to explain why it's better to use the `<>` version. For example:
```suggestion
In contrast, using `coroutine_handle<>` increases the reusability of the code because this parameter type supports all promise types.
```

> This rule raises an issue for the implementation of `await_suspend` that accepts handles to a specific promise and does not use that information in implementation.

Reviewer 2: ```suggestion
This rule raises an issue for the implementation of `await_suspend` that accepts handles to a specific promise type and yet does not use that information.
```

   /* ... */
   bool await_suspend(std::coroutine_handle<Promise> current) { // Noncompliant
     return event.register_callback([current] {
              current.resume();
            });     
   }
};

struct Awaiter2
{
   Event& event;
   /* ... */
   bool await_suspend(std::coroutine_handle<PromiseA> current) { // Noncompliant
     return event.register_callback([current] {
              current.resume();
            });     
   }
   bool await_suspend(std::coroutine_handle<PromiseB> current) { // Noncompliant
     return event.register_callback([current] {
              current.resume();
            });     
   }
};

struct Awaiter3
{
   Event& event;
   /* ... */
   template<typename PromiseType>
   bool await_suspend(std::coroutine_handle<PromiseType> current) { // Noncompliant
     return event.register_callback([current] {
              current.resume();
            });     
   }
};

struct AwaiterUsingPromise
{
  /* ... */
  void await_suspend(std::coroutine_handle<Promise> current) { // Compliant
    auto wokeUpTime = std::chrono::system_clock::now() + std::chrono::seconds(10);
    current.promise().executor().schedule_at(wokeUpTime, current); // promise used here
  }
  

> };

Reviewer 2: I wonder if this should go in the "Compliant Solution" section?

----

== Compliant Solution

----

> struct Awaiter

Reviewer 2: A tiny suggestion to emphasize the point of the rule:
```suggestion
struct GenericAwaiter
```

Reviewer 3: Added approparite comment
